pretyping/pretyping.ml

```diff
@@ -922,17 +979,80 @@ struct
+      let var',env' = push_rel ~hypnaming sigma var env in
+      let sigma, j' = eval_pretyper self ~program_mode ~poly resolve_tc rng env' sigma c2 in
+      let name = get_name var' in
+      let resj = judge_of_abstraction !!env (orelse_name name name'.binder_name) j j' in
+      discard_trace @@ inh_conv_coerce_to_tycon ?loc ~program_mode resolve_tc env sigma resj tycon
+    with
+    | PretypeError _ as pte ->
+      let reraise = Exninfo.capture pte in
+
+      (* Maybe tycon' is a primitive projection? *)
+
+      match tycon' with
+      | None ->
+        Exninfo.iraise reraise
+      | Some ty ->
+        (* We should really re-use the reduction performed in split_tycon *)
+        let t = Reductionops.whd_all !!env sigma ty in
+        let head, extra_args = EConstr.decompose_app sigma t in
+        match EConstr.kind sigma head with
+        | Proj (proj, arg) when isEvar sigma arg ->
+          begin
+            try
+              (* Quick check to see if this can succeed at all, avoiding a costly unification call if possible. *)
+              let _ = Recordops.lookup_canonical_conversion (Names.GlobRef.ConstRef (Projection.constant proj), Recordops.Prod_cs) in
+              let sigma, domain_hole = new_type_evar env sigma loc in
+              let sigma, codomain_hole = new_type_evar env sigma loc in
+              let cs_type = mkProd (Context.anonR, domain_hole, codomain_hole) in
+
+              let sigma = Evarconv.unify
+                  !!env
+                  sigma
+                  Reduction.CUMUL
+                  (EConstr.mkApp (head, Array.of_list extra_args))
+                  cs_type
+              in
+              pretype_lambda self (name, bk, c1, c2) ?loc ~program_mode ~poly resolve_tc (Some cs_type) env sigma
+            with
+            | Not_found | UnableToUnify _ -> Exninfo.iraise reraise
+            end
+        | _ ->
+          try
+            let (proj, evar, params) = Recordops.decompose_reduced_canonical_projection sigma head in
+
+            (* Quick check to see if this can succeed at all, avoiding a costly unification call if possible. *)
+            let _ = Recordops.lookup_canonical_conversion (Names.GlobRef.ConstRef proj, Recordops.Prod_cs) in
+
+            (* Feedback.msg_debug (str "pretyp:" ++ Constr.debug_print (EConstr.Unsafe.to_constr evar)); *)
+            let args = params in
+            let args = Array.of_list (args @ [evar] @ extra_args) in
+
+            (* let sigma, ty_hole = new_type_evar env sigma loc in *)
+            (* codomain  *)
+            let sigma, domain_hole = new_type_evar env sigma loc in
+            let sigma, codomain_hole = new_type_evar env sigma loc in
+            let cs_type = mkProd (Context.anonR, domain_hole, codomain_hole) in
+
+            let sigma = Evarconv.unify
+                !!env
+                sigma
+                Reduction.CUMUL
+                (mkApp (mkConst proj, args))
```

Comment:
mkConst (non U) is certainly incorrect
You may be able to get universes out of decompose_reduced_canonical_projection, it should be the same instance as the instance for the inductive type.

Reply:
Should be fixed now.